### DISCUSSION TITLE
JavetExecutionException: Error: Cannot find module

### DISCUSSION THREAD

**Discussion by mrinalsharma** [caoccao/Javet#99] [Q&A]
Can I write my own node modules and run them through Javet. In this example, demo.js require component.js
My code looks like this:

```			
                ClassLoader classLoader = getClass().getClassLoader();
		URL resource = classLoader.getResource("nodejs/components/demo.js");
               try (JavetEnginePool<NodeRuntime> javetEnginePool = new JavetEnginePool<NodeRuntime>()) {
			    javetEnginePool.getConfig().setJSRuntimeType(JSRuntimeType.Node);
			    try (IJavetEngine<NodeRuntime> iJavetEngine = javetEnginePool.getEngine()) {
			        NodeRuntime nodeRuntime = iJavetEngine.getV8Runtime();
			        V8ValueObject v8ValueObject = nodeRuntime.createV8ValueObject();
			        nodeRuntime.getGlobalObject().set("NodeJs", v8ValueObject);
			        nodeRuntime.getGlobalObject().set("console", v8ValueObject);
			        v8ValueObject.bind(this);
			        System.out.println(nodeRuntime.getExecutor(new File(resource.toURI())).executeString());
			    }
			}```

When I run this code I see the following error. The module is stored in "resources\nodejs\components\node_modules" folder.
```WARNING: 2 V8 callback context object(s) not recycled.
Exception in thread "main" com.caoccao.javet.exceptions.JavetExecutionException: Error: Cannot find module 'Component'
Require stack:```

**Comment by caoccao** [comment]
Your code looks a little bit weird to me.

# Node.js is Not Able to Locate `resources` in Jar File

I'm not sure if you package all the scripts in the jar file. If that's YES, you need to unpack them to a physical location for Node.js to access.

# Locate the `node_modules`

Node.js locates `node_modules` based on the following.

* `require()` Root
* Working Directory
* `__dirname`
* `__filename`

I suggest you to write a simple script printing these env variables to verify the runtime env is correct. The best practice is to put the entry script in a dedicated folder which stays with `node_modules` side by side. Node.js usually searchs from `dedicated_folder/../` for `node_modules`.

Please refer to this [doc](https://www.caoccao.com/Javet/reference/resource_management/modularization.html) for the official recommendation in this case.

**Comment by mrinalsharma** [comment]
Does Javet keep the cache of Node modules?  I am changing the code of NodeJS module but Javet continues to run old code.

**Reply by caoccao** [reply]
> Does Javet keep the cache of Node modules? I am changing the code of NodeJS module but Javet continues to run old code.

Yes, Javet caches the compiled modules in the memory. Here are 3 typical ways of invalidating the cache.

1. **resetContext()** - It's the most recommended way.
2. **removeV8Module()** - Removing a particular module explicitly gives application flexibility.
3. **createV8Runtime()** - Just create a new V8 runtime. It's cheap.

**Comment by mrinalsharma** [comment]
I am seeing this issue after application restart so there should not be any code in memory. Does the engine cache on disk?

**Reply by caoccao** [reply]
> I am seeing this issue after application restart so there should not be any code in memory. Does the engine cache on disk?

No, everything is in the memory. Now, I think your `node_modules` might be polluted. I'm not a big fan of the way NPM manages things. There could be many `node_modules` folders distributed in many locations. Please verify it's the right `node_modules` you want Javet Node.js mode to locate.

**Reply by mrinalsharma** [reply]
Javet's getExecutor() method sets the require folder as a parent of the script path, so I am assuming that all the node modules  should be loaded from the parent folder.

**Reply by caoccao** [reply]
Sure, you may change the path after `getExecutor()`.

I think an easy way is to put some `console.log()` in your modules to find out which one is loaded.

**Comment by mrinalsharma** [comment]
I was able to resolve the issue by giving absolute path of my Node module in require().

example.
```
var http = require('http');
var dt = require('C:\\Users\\admin\\Component\\MYComponent');
```
